pkg/controller/volume/persistentvolume/pv_controller_base.go

```diff
@@ -368,17 +372,70 @@ func (ctrl *PersistentVolumeController) updateVolumeMigrationAnnotationsAndFinal
 	return newVol, nil
 }
 
-// updateMigrationAnnotationsAndFinalizers takes an Annotations map and finalizers slice and checks for a
-// provisioner name depending if the annotation came from a PVC or not.
-// It will then add a "pv.kubernetes.io/migrated-to" annotation if migration with
-// the CSI driver name for that provisioner is "on" based on feature flags, it will also
-// remove the PV deletion protection finalizer and "pv.kubernetes.io/migrated-to" annotation
-// if migration is "off" for that provisioner in rollback scenarios. Returns true if the annotations
-// map or finalizers were modified and false otherwise.
-// Parameters:
-// - claim: true means the ann came from a PVC, false means the ann came from a PV
-func updateMigrationAnnotationsAndFinalizers(cmpm CSIMigratedPluginManager, translator CSINameTranslator,
-	ann map[string]string, finalizers *[]string, claim bool) bool {
+// modifyDeletionFinalizers updates the finalizers based on the reclaim policy and if it is a in-tree volume or not.
+// The in-tree PV deletion protection finalizer is only added if the reclaimPolicy associated with the PV is `Delete`.
+// The in-tree PV deletion protection finalizer is removed if the reclaimPolicy associated with the PV is `Retain` or
+// `Recycle`, removing the finalizer is necessary to reflect the recalimPolicy updates on the PV.
+// The method also removes any external PV Deletion Protection finalizers added on the PV, this represents CSI migration
+// rollback/disable scenarios.
+func modifyDeletionFinalizers(cmpm CSIMigratedPluginManager, volume *v1.PersistentVolume) ([]string, bool) {
+	modified := false
+	var inFinalizers []string
+	if !utilfeature.DefaultFeatureGate.Enabled(features.HonorPVReclaimPolicy) {
+		return inFinalizers, false
+	}
+	if volume.Finalizers != nil {
+		inFinalizers = append(inFinalizers, volume.Finalizers...)
+	}
+	if !metav1.HasAnnotation(volume.ObjectMeta, storagehelpers.AnnDynamicallyProvisioned) {
+		// PV deletion protection finalizer is currently supported only for dynamically
+		// provisioned volumes.
+		return inFinalizers, false
+	}
+	provisioner := volume.Annotations[storagehelpers.AnnDynamicallyProvisioned]
```

Comment:
What if there is no such annotation? It will fall under `Migration disabled` below, which is confusing.

Reply:
```
if !metav1.HasAnnotation(volume.ObjectMeta, storagehelpers.AnnDynamicallyProvisioned) {
		// PV deletion protection finalizer is currently supported only for dynamically
		// provisioned volumes.
		return volume.Finalizers, false
	}
```

The above would return if there is no such annotation. But, I guess the key can exist without value. In which case we check if the provisioner is an in-tree volume, if not we return.


Got it, I guess having a specific block for `Migration disabled` is a bit confusing. I think a better approach to make the code clear is we treat `Migration enabled` as a special case, i.e, if enabled we remove the in-tree pv deletion protection finalizer and return immediately.